lib/rmg/CMapGenerator.cpp

```diff
@@ -285,17 +285,50 @@ void CMapGenerator::genZones()
 	for(const auto & option : tmpl->getZones())
 	{
 		auto zone = std::make_shared<CRmgTemplateZone>(this);
-		zone->setOptions(option.second.get());
+		zone->setOptions(*option.second.get());
 		zones[zone->getId()] = zone;
 	}
 
 	CZonePlacer placer(this);
 	placer.placeZones(&rand);
 	placer.assignZones();
+	
+	//add special zone for water
+	zoneWater.first = zones.size()+1;
+	zoneWater.second = std::make_shared<CRmgTemplateZone>(this);
+	{
+		rmg::ZoneOptions options;
+		options.setId(zoneWater.first);
+		options.setType(ETemplateZoneType::WATER);
+		zoneWater.second->setOptions(options);
+	}
 
 	logGlobal->info("Zones generated successfully");
 }
 
+void CMapGenerator::createWaterTreasures()
+{
+	ui32 waterTreasure = getZoneWater().second->getTileInfo().size();
+	//add treasures on water
+	getZoneWater().second->addTreasureInfo(CTreasureInfo{1000, 3000, 9});
```

Comment:
What is 9 here? Amount of treasure items? Should it depend on map size?

Reply:
It's treasure density, and it's properly scaled afterwards.

Reply:
Hey, got it: https://forum.df2.ru/index.php?showtopic=32286

> Changes to the treasures of the water zone.
> How it was.
> Value 100-1000: Frequency of occurence 5.
> Value 2000-6000: Frequency of occurence 1.
> 
> How it is now.
> Value 100-1000: Frequency of occurence 6.
> Value 1000-4000: Frequency of occurence 3.
> Value 4000-12000: Frequency of occurence 1.